### PR TITLE
Products with qtybook

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/11177

### CODE REVIEW

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                      LEFT OUTER JOIN M_Inventory inv ON inv.m_inventory_id = il.m_inventory_id
+                      LEFT OUTER JOIN M_MovementLine ml ON ml.m_movementline_id = cd.m_movementline_id
+                      LEFT OUTER JOIN M_Locator ml_locFrom ON ml.m_locator_id = ml_locFrom.m_locator_id
+                      LEFT OUTER JOIN M_Locator ml_locTo ON ml.m_locatorto_id = ml_locTo.m_locator_id
+                      LEFT OUTER JOIN M_Movement m ON m.m_movement_id = ml.m_movement_ID
+                      LEFT OUTER JOIN m_inoutline iol ON iol.m_inoutline_id = cd.m_inoutline_id
+                      LEFT OUTER JOIN m_inout io ON iol.m_inout_id = io.m_inout_id
+             WHERE TRUE
+               AND cd.m_costelement_id = p_M_CostElement_ID
+               AND COALESCE(cd.m_inoutline_id, cd.m_movementline_id, cd.m_inventoryline_id) IS NOT NULL
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql:52]: pls, filter by parameters as soon as possible (i.e. in this CTE),
in order to reduce the amount of data processed as soon as posible.

Comment (teosarca) [reply backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql:52]: i.e. filter by product, filter by warehouse

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+         )
+
+
+SELECT wh.name  as WarehouseName,
+       p.value  as ProductValue,
+       p.name   as ProductName,
+
+       sum(qty) as qtyBook
+from m_product p
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql:61]: why `m_product outer join tmp_costdetails`?
is there any reason to show ALL products?

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql (hunk 2 above)
Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql:61]: No. Fixed.

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql (hunk 1 above)
Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql:52]: ✔️

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql [hunk 3]
```diff
@@ -50,30 +49,37 @@ WITH tmp_costdetails AS
              WHERE TRUE
                AND cd.m_costelement_id = p_M_CostElement_ID
                AND COALESCE(cd.m_inoutline_id, cd.m_movementline_id, cd.m_inventoryline_id) IS NOT NULL
+               AND (p_M_Product_ID <= 0 OR p_M_Product_ID = cd.m_product_id)
+               AND (
+                     (p_date IS NULL AND COALESCE(inv.movementdate, io.MovementDate, m.movementdate) <= NOW())
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql]: but why enforcing NOW()?

Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/report_M_Product_QtyBooked_For_Date.sql]: Right. The way you suggested in your proto is better than mine so I replaced. Please, refresh

**Comment by metas-rc** [comment]
Approved by Teo.

**Pull request merged**